File: src/main/java/it/facile/form/EditTextOnSubscribe.java
package it.facile.form;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import rx.Observable;
import rx.Subscriber;
import rx.android.MainThreadSubscription;

public class EditTextOnSubscribe implements Observable.OnSubscribe<CharSequence> {

    private final EditText editText;
    private boolean initialVal;

    public EditTextOnSubscribe(EditText editText, boolean initialVal) {
        this.editText = editText;
        this.initialVal = initialVal;
    }

    @Override
    public void call(final Subscriber<? super CharSequence> subscriber) {

        final TextWatcher watcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // If the subscriber is subscribed emit the new text value
                if (!subscriber.isUnsubscribed()) {
                    subscriber.onNext(s);
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        };
        editText.addTextChangedListener(watcher);

        // Add the a main thread safe subscription
        subscriber.add(new MainThreadSubscription() {
            @Override
            protected void onUnsubscribe() {
                editText.removeTextChangedListener(watcher);
            }
        });

        if (initialVal) subscriber.onNext(editText.getText());
    }
}

File: src/main/kotlin/it/facile/form/ui/utils/EditTextOnSubscribe.kt
package it.facile.form.ui.utils

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import rx.Observable
import rx.Subscriber
import rx.android.MainThreadSubscription

class EditTextOnSubscribe(private val editText: EditText, private val initialVal: Boolean, val nonObservedChanges: ((EditText) -> Unit)?)
    : Observable.OnSubscribe<CharSequence> {

    override fun call(subscriber: Subscriber<in CharSequence>) {

        val watcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (subscriber.isUnsubscribed) return

                subscriber.onNext(editText.text)
                editText.removeTextChangedListener(this)
                nonObservedChanges?.invoke(editText)
                editText.addTextChangedListener(this)
            }

            override fun afterTextChanged(s: Editable) {

            }
        }
        editText.addTextChangedListener(watcher)

        // Add the a main thread safe subscription
        subscriber.add(
                object : MainThreadSubscription() {
                    override fun onUnsubscribe() {
                        editText.removeTextChangedListener(watcher)
                    }
                })

        if (initialVal) subscriber.onNext(editText.text)


    }
}
